web-gateway/src/main/java/controllers/AbstractController.java

```diff
@@ -45,19 +48,14 @@ public AbstractController(MessagesApi messagesApi, UserService userService) {
     protected CompletionStage<Nav> loadNav(UUID userId) {
         return loadNav(Optional.of(userId));
     }
-
-    protected CompletionStage<Nav> loadNav(Optional<UUID> userId) {
-        return userService.getUsers(Optional.of(0), Optional.of(10)).invoke().thenApply(users -> {
-            Optional<User> currentUser = userId.flatMap(id -> {
-                for (User u: users.getItems()) {
-                    if (u.getId().equals(id)) {
-                        return Optional.of(u);
-                    }
-                }
-                return Optional.empty();
-            });
-            return new Nav(messagesApi.preferred(Collections.emptyList()), users.getItems(), currentUser);
-        });
-    }
-
-}
+    protected CompletionStage<Nav> loadNav(Optional<UUID> userId) { CompletionStage<Optional<User>> u;
+    if (userId.isPresent()) {
+        u = userService.getUser(userId.get())
+                .handleRequestHeader(authenticate(userId.get()))
+                .invoke()
+                .thenApply(resp -> Optional.of(resp)); } else {
+        u = CompletableFuture.completedFuture(Optional.empty()); }
+         CompletionStage<PaginatedSequence<User>> users = userService.getUsers(Optional.of(0), Optional.of(10)).invoke();
+    return users.thenCombineAsync( u, (userPaginatedSequence, maybeLoggedUser) -> new Nav(messagesApi.preferred(Collections.emptyList()), userPaginatedSequence.getItems(), maybeLoggedUser)); }
+
```

Comment:
I would also extract the code that creates the `Nav` to a separate method that doesn't deal with any `CompletionStage`:

```java
    private Nav getNav(PaginatedSequence<User> users, Optional<User> currentUser) {
        return new Nav(messagesApi.preferred(Collections.emptyList()), users.getItems(), currentUser);
    }
```

Then, you can use it with a method reference rather than a lambda:

```java
return users.thenCombineAsync(u, this::getNav);
```

This has two benefits:

1. You don't need to have two different names for the same thing (`users` & `userPaginatedSequence`, `u` or `currentUser` & `maybeLoggedUser`)
2. Each method stays focused on a single level of abstraction and you don't have to mix `CompletionStage`-handling code with value-handling code